### ISSUE TITLE
Proposal: Create option to disable `no-shadow` rule for switch/case statements

### ISSUE THREAD

**Issue by guerrero** [eslint/eslint#4473]
It would be nice to create an option for `no-shadow` rule to prevent errors when variables are declared in different `case` statements. For example, the following code returns an error and that's OK due that I'm creating a new scope but, as far as I know, there isn't a real shadowing problem here because these `case`s are mutually exclusive
```js
function foobarbaz(foo) {
  switch (foo) {
    case 'bar':
      const foobar = 'foobar'
      return foobar
    case 'baz':
      return function(foobar) { // error: 'foobar' is already declared in the upper scope
        console.log('baz')
      }
  }
}
```
**1. When the rule won't warn enabling this option**
In a code similiar to the previous example and a code similar to the following:
```js
function foobarbaz(foo) {
  switch (foo) {
    case 'bar':
      const foobar = 'foobar'
      return foobar
    case 'baz':
      return function(bar) {
        const foobar = true // error: 'foobar' is already declared in the upper scope
        console.log('baz')
      }
  }
}
```
**2. Whether the rule prevents an error or is stylistic.**
This option rule prevents an eslint error when there's no real problem

**3. Why the option should be in the core instead of creating a custom rule.**
Because is related to a core rule

Something like `no-shadow: [2, { "disableOnSwitch": "true" }]` would be helpful in this case. Also, if you're interested, I will try to help creating a PR.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
If I'm not mistaken, your assumption is incorrect (sorry, I don't use ES6, so I might be wrong here). `case` statements don't create code blocks, so all of the declarations inside case statements are scoped to the `switch`, in which case this rule is doing the right thing. See #4278 for more information.

**Comment by michaelficarra** [comment]
@ilyavolodin is correct here, the switch cases share a single lexical environment.

**Comment by guerrero** [comment]
Yes @ilyavolodin, `case` statements don't create code blocks – es6 has the same behaviour – and I'm not trying to say that, sorry if my explanation is not sufficiently clear.

I don't want a feature to enable something like a scope for `case`s, instead what I intend to propose is an option to enable `no-shadow` on `switch` statements due that its `case`s are mutually exclusive when there's a `return` or `break` and therefore we already know that only one statement block will be executed. From MDN:

> A switch statement first evaluates its expression. It then looks for the first case clause whose expression evaluates to the same value as the result of the input expression (using strict comparison, ===) and transfers control to that clause, executing the associated statements. (If multiple cases match the provided value, the first case that matches is selected, even if the cases are not equal to each other.)

**Comment by ilyavolodin** [comment]
I'm not really sure it's a good idea to create an option like that, since it's actually likely to cause side effects. Also, this would also only apply to cases which are terminated by `return` or `break`, and until we have full code analysis in place, we can't create rules that will guarantee that every branch of the code is terminated.

**Comment by nzakas** [comment]
@guerrero your first example is behaving correctly. The `const` declaration applies to the whole `switch`, not just the `case`. Therefore, your function is shadowing a variable from the enclosing scope, so I don't see us adding an exception for this.
